### ISSUE TITLE
Search pre-fill?

### ISSUE THREAD

**Issue by chrisspen** [CloudCannon/pagefind#144]
How would you pre-fill the Pagefind search bar?

I'm trying to implement a simple search term pre-fill where a "q" parameter is extracted from the URL's querystring and used to pre-fill the Pagefind search field so I can provide a link to a page with "?q=some+search+term" and have Pagefind automatically begin a search once the page loads.

However, I'm having trouble finding a way to access the auto-generated Pagefind object to trigger the search. I can fill in the auto-generated search bar, but Pagefind seems to ignore any key events generated by the form field.

For example, if my Pagefind object has the id "search", and I try:

    $('#search [type=text]').val('some search terms');
    $('#search [type=text]')[0].dispatchEvent(new KeyboardEvent('keydown', { key: "enter" }));

to simulate the user typing in the search terms and pressing enter, the search bar gets populated with the text, but the search never gets executed.

What's the proper way to dispatch events to the Pagefind object?

**Comment by bglw** [comment]
I haven't documented any methods on the object returned from `new PagefindUI` yet — but one does exist for this (and is used in the tests, so it won't go anywhere):

```js
let pagefind = new PagefindUI({ element: "#search" });
pagefind.triggerSearch("some search terms");
```

Let me know if that solves it for ya

**Comment by chrisspen** [comment]
Perfect. Thanks.

**Comment by chrisspen** [comment]
Is there any reason why `triggerSearch()` would return different results than entering the same search text in manually?

If I enter a term manually, Pagefind returns 1733 results, but if I enter the same term via `triggerSearch()` it only returns 38 results.

I'm not sure how long it takes Pagefind to initialize behind the scenes of I'm calling `triggerSearch()` via a setTimeout like:

    window.addEventListener('DOMContentLoaded', (event) => {
    
        pagefind = new PagefindUI({
            element: "#search",
            baseUrl: "/",
            debounceTimeoutMs: 500,
            showImages: false,
            mergeIndex: [
                {bundlePath: "/static/myindex"}
            ]
        });

        const urlParams = new URLSearchParams(window.location.search);
        const q = urlParams.get('q');
        if(q){
            setTimeout(function(){
                console.log('Searching:'+q);
                pagefind.triggerSearch(q);
            }, 1000);
        }

    });

**Comment by bglw** [comment]
Hmm, I can't think of a strong reason why that would happen. The only difference between the two would be as you're typing the search query more indexes would be preloaded. 

What behaviour do you see if you paste the search term in, versus typing it in? 

(Also, no need to wait to call triggerSearch, Pagefind will delay internally until it has fully initialized)

**Comment by chrisspen** [comment]
Both seem accurate, but the manual typing just finds more results.

Yeah, I suspected it was from index preloading. I didn't show it in my sample code, but I'm using about 10 index paths in my mergeIndex list.

If Pagefind is waiting until it's fully initialized, why would the triggerSearch return fewer results than the manual search? Shouldn't they return the identical results? How do I know when Pagefind has loaded all indexes in the mergeIndex list?

**Comment by bglw** [comment]
Hmm, if that is indeed the cause then it is a bug. Looking into it

**Comment by bglw** [comment]
Found, fixed, [released in v0.10.0](https://github.com/CloudCannon/pagefind/releases/tag/v0.10.0) — let me know if you're still seeing an issue.

**Comment by chrisspen** [comment]
Behavior seems to be the same. Other than reinstalling in my project root with `npx pagefind@v0.10.0 --source "public"` and deleting and rebuilding all my indexes, is there anything else I need to do to ensure it's using the new version?

**Comment by bglw** [comment]
Hi @chrisspen — did things end up working out for you or should I dig deeper?

**Comment by chrisspen** [comment]
Yeah, results still differ slightly, even with the new version. Can't find any issue on my end. Not a big deal though.

**Comment by bglw** [comment]
Closing out as [v1.5.0](https://github.com/Pagefind/pagefind/releases/tag/v1.5.0) shipped a bunch of stuff in this area, so I strongly suspect this has been fixed. Do re-open if it's still seen on the latest Pagefind! (or preferably a new issue for maximum noise 😄)
